develop/tutorials/articles/140-application-security/01-application-permissions/03-registering-permissions.markdown

```diff
@@ -49,6 +53,77 @@ or group template (site template). These are used internally for those objects,
 so you'd only use them if you were customizing functionality for creating and
 deleting them. 
 
+Now you're ready to register your entities with the permissions service. 
+
+## Registering Entities to the Permissions Service
+
+The permissions service that's running needs to know about your entities and how
+to check permissions for them. This requires creating a permissions registrar
+class. 
+
+1.  In your service bundle, create a package that by convention ends in
+    `internal.security.permission.resource`. For example, the Blogs
+    application's package is named
+    `com.liferay.blogs.internal.security.permission.resource`. 
+
+2.  Create a class in this package called `[Entity
+    Name]ModelResourcePermissionRegistrar`. For example, the Blogs application's
+    class is named `BlogsEntryModelResourcePermissionRegistrar`. 
+
+3.  This class is a component class that requires overriding the `activate`
+    method to register the permissions logic you want for your entities. For
+    example, this is how the Blogs application registers its permissions: 
+
+        @Component(immediate = true)
+        public class BlogsEntryModelResourcePermissionRegistrar {
+
+            @Activate
+            public void activate(BundleContext bundleContext) {
+                Dictionary<String, Object> properties = new HashMapDictionary<>();
+
+                properties.put("model.class.name", BlogsEntry.class.getName());
+
+                _serviceRegistration = bundleContext.registerService(
+                    ModelResourcePermission.class,
+                    ModelResourcePermissionFactory.create(
```

Comment:
They can optionally implement ModelResourcePermission directly as a component as it has less requirements to using the ModelResourcePermissionFactory.  However, using the factory means they get to have a per request permission cache applied to their permission checks for free.

Reply:
@Preston-Crary I was trying to keep it as simple as possible by showing the best practice. Your doc draft seemed to indicate that this was the best practice, rather than implementing it directly for the reason you indicated (request permission cache). 

Under what circumstances would developers choose to implement ModelResourcePermission directly? I think it's best to tell developers the reasons why an option is available, rather than just tell them all the options.

Reply:
The option exists because it's necessary if their model does not implement GroupedModel.  This is determined by which fields (columns) they created their model with in their service.xml.  Implementing it directly allows for any ServiceBuilder model to implement permissions.

Reply:
And non-Service Builder models? Like, what if somebody is using JPA or Hibernate directly? Can they still use our permissions system?

Reply:
They would need to implement ClasssedModel, I'm not sure if they would need anything else.